Replace lodash helpers

askmatey opened the pull request:
- :heavy_minus_sign: ~~Add a string formatter~~
- :heavy_plus_sign: Register a the [camelcase](https://www.npmjs.com/package/camelcase) package to package.json
- :heavy_minus_sign: ~~Remove and replace `lodash.camelcase` with format.camelCase method~~
- Replace lodash.camelcase with the `camelcase` string formatter
- :heavy_minus_sign: ~~Remove and replace `lodash.kebabcase` with format.kebabCase method~~
- :heavy_plus_sign: Add a new `kebabCase` string formatter
- Replace `lodash.kebabcase` with the `kebabCase` string formatter
- Add a new `isPlainObject` utility
- Replace `lodash.isplainobject` with the `isPlainObject` utility
- :heavy_minus_sign: ~~Add a unit test for the string formatters~~
- :heavy_plus_sign: Add a new unit test for for the `camelCase` string formatter
- Add a new unit test for the `isPlainobject` utility
- Unregister the deprecated lodash depencencies from package.json

garbles commented:
Thanks for this! Looking pretty good. I do have a few comments coming up.

garbles commented on src/format.js:
```diff
@@ -0,0 +1,15 @@
+      return index === 0 ? letter.toLowerCase() : `-` + letter.toLowerCase()
+    })
+  },
+}
```
Would prefer if these were their own modules.

```js
module.exports = function camelCase (str) { ... }
```

```js
module.exports = function kebabCase (str) { ... }
```

askmatey replied:
Another option would be, to outsource the regex string into an constant like this

```js
const regex = /(?:^\w|[A-Z]|\b\w|\s+)/g

module.exports = {
  camelCase: function camelCase (str) {
    return str.replace(regex, (letter, index) => {
      if (+letter === 0) return ``
      return index === 0 ? letter.toLowerCase() : letter.toUpperCase()
    })
  },

  kebabCase: function kebabCase (str) {
    return str.replace(regex, (letter, index) => {
      if (+letter === 0) return ``
      return index === 0 ? letter.toLowerCase() : `-` + letter.toLowerCase()
    })
  },
}
```

garbles replied:
That's no problem. I'm just pretty big on making sure that private modules do a single thing.

garbles commented on src/isPlainObject.js:
```diff
@@ -0,0 +1,20 @@
+const fnToString = (fn) => Function.prototype.toString.call(fn)
```
There is already a module for this. https://github.com/garbles/yolk/blob/master/src/isFunction.js

askmatey replied:
:+1:

askmatey replied:
Nah, do not think so

garbles commented on test/unit/formatString-test.js:
```diff
@@ -0,0 +1,25 @@
+
+  t.equal(format.camelCase(`foo bar`), resultString)
+  t.equal(format.camelCase(`--foo-bar`), resultString)
+  t.equal(format.camelCase(`__foo_bar__`), resultString)
```
Can you also test cases with a single word, e.g. `foobar => foobar` and with more than two cases, `foo-bar-baz__ => fooBarBaz`

askmatey replied:
Yes

garbles commented on test/unit/formatString-test.js:
```diff
@@ -0,0 +1,25 @@
+  t.equal(format.kebabCase(`fooBar`), resultString)
+  t.equal(format.kebabCase(`foo bar`), resultString)
+  t.equal(format.kebabCase(`--foo-bar`), resultString)
+  t.equal(format.kebabCase(`__foo_bar__`), resultString)
```
Same with this one. `foo` and `foo-bar-baz`

garbles commented on test/unit/isPlainObject-test.js:
```diff
@@ -0,0 +1,18 @@
+  t.equal(isPlainObject([ 1, 2, 3 ]), false)
+  t.equal(isPlainObject(null), false)
+  t.equal(isPlainObject(), false)
+  t.equal(isPlainObject({ 'foo': 1, 'bar': 2 }), true)
```
Please also test boolean and string values.

askmatey replied:
Yes

askmatey commented:
The travis checks are somehow failing, I am working on that

askmatey commented:
Changes

- Added the [camelcase](https://www.npmjs.com/package/camelcase) package (MIT) to package.json
- Removed the previous formatter
- Outsourced the kebabCase formatter to a single file (`kebabCase.js`)
- Updated the unit tests

Status
The kebabCase and isPlainObject unit tests are passing. However, I am stuck at that `UnexpectedVirtualElement` error, see [Issue 48: Test suite times out a lot](#48)

I have created a bundle locally and tested it successfully with my app.

askmatey commented:
I have added another `isComponent`check to the chain, since CompositeComponents are defacto *plain objects*

```sh
Object { 
  name: "YolkCompositeComponent_Button", 
  id: "Button", 
  _fn: Button(), 
  _props: Object, 
  _children: Array, 
  _component: Object, 
  _props$: Object, 
  _children$: Object,
  __proto__: {
    destroy: destroy(),
    init: init(),
    type: 'Widget',
    update: update()
  }
}
```

garbles commented:
> since CompositeComponents are defacto plain objects

lodash `isPlainObject` says otherwise
